[Step3] 날씨의 아이

July911 opened the pull request:
안녕하세요 나무 !!! 
날씨의 아이 Step3가 끝나서 PR드립니다 !

비동기처리하는 부분을 대부분 RxSwift를 사용한 Observable로 만들어서 리팩토링해보았습니다 !
나무가 RxSwift를 잘하셔서 제가 부족한 부분을 많이 지적해주실 수 있으실 것 같아요. 
## 구현한 부분 
### Reactive Extension 으로 필요한 부분 구현 
```swift 
private extension Reactive where Base: UIImageView {
    func loadCacheView(webView: WKWebView) -> Binder<UIImage> {
        return Binder(self.base) { ImageView, image in
            webView.isHidden = true
            base.isHidden = false
            ImageView.image = image
        }
    }
}`
``` 


### Observable 병합을 통한 비지니스 로직 이동 
```swift
 func transform(input: Input) -> Output {
        let entireCities = self.searchUseCase.extractCities().asObservable()
        let filteredCities = input.searchBarText.distinctUntilChanged()
            .flatMap { (text) -> Observable<[City]> in
            return self.searchUseCase.search(text ?? "")
        }

        let combined = Observable.combineLatest(entireCities, filteredCities) { (city, fil) -> [City] in
            return fil.isEmpty ? city : fil
        }.asDriver(onErrorJustReturn: [])
        
     
     
        input.didSelectedCell.subscribe(onNext: { (city) in
            self.coordinator.occuredViewEvent(with: .presentDetailShowUIViewController(cityName: city))
        }).disposed(by: self.bag)
        
        return Output(initialCities: combined)
    }
``` 

### Input & Output Modeling 

```swift
  struct Input {
        let viewWillAppear: Observable<Void>
        let didSelectedCell: Observable<City>
        let searchBarText: Observable<String?>
    }
    
    struct Output {
        let initialCities: Driver<[City]>
    }
``` 
Transform 메서드가 인터페이스가 되어서 ViewController가 이 메서드만 호출하고 그로 인해 바인딩이 되게 구현했습니다.

### 함수형 프로그래밍 구현

```swift
 func search(name: String?) -> Observable<[City]> {
        
        guard let name = name, name != ""
        else {
            self.fetchCityList()
            return self.assetData.asObservable()
        }
        let filteredCity = assetData.value.filter { $0.name.hasPrefix(name) }
        
        return Observable<[City]>.just(filteredCity)
    }
```

SearchBar를 통해 search를 할 경우에 원본 데이터는 손상시키지않고 언제나 같은 값이 나오도록 함수형 프로그래밍으로 메서드를 구현해보았습니다!

## 궁금한점 


### ViewController의 로직 

```swift 
self.navigationItem.searchController?.searchBar.rx.text
            .distinctUntilChanged()
            .subscribe(onNext: { event in
                self.listTableView.reloadData()
            }).disposed(by: self.bag)
        
        self.listTableView.rx.itemSelected
            .withUnretained(self)
            .subscribe(onNext: { (self,index) in
            self.listTableView.deselectRow(at: index, animated: true)
            }).disposed(by: self.bag)
        
``` 

SearchBar의 Text가 들어왔을 때 reloadData() 
item이 seleted 되었을 때  deseletRow되는 로직이 ViewController 에 들어있는데 이 정도까지는 괜찮을지가 궁금합니다 !

July911 committed:
feat: RxSwift SPM으로 import

July911 committed:
fix: libXCTestSwiftSupport 버그 수정

- RxTest 제거
- Enable Testing Search Paths 를 Yes 로 변경

July911 committed:
feat: APICaller 객체 RxSwift로 수정

- Escaping Closure 대신 Observable이 리턴되게 수정

July911 committed:
feat: WeatherRepository RxSwift 적용

- APIService의 리턴값을 Single로 변경
- WeatherRepository에서 Domain의 model인 Observable<TodayWeather>로 변경

July911 committed:
feat: CitySearchRepository RxSwift 적용

- 원본이 되는 데이터 assetData를 BehaviorRelay로 변경
- 값을 search했을 때 accept을 사용하여 값을 변경

July911 committed:
feat: UseCase RxSwift 적용

- 각 도시의 오늘날씨를 Observable로 받게 변경
- Repository의 메서드들의 return값을 제거하여 함수형프로그래밍으로 변경
- 메인 list 화면의 뷰를 BehaviorRelay의 시퀀스를 따르게 하여 뷰의 정보가 끊이지 않게 구현

July911 committed:
feat: ViewModel Protocol에 Input&Output 모델링 정의

- ViewModel 에 nestedType으로 Struct Input,Output 생성
- ViewController가 추후 사용할 transform 메서드 구현

July911 committed:
refactor: Input&Output 모델링을 위한 ViewModel 추상화 제거

- associated 타입으로 정의한 protocol은 타입으로서 사용 불가
- transform 메서드의 ViewController 에서 사용을 위한 추상화 제거

July911 committed:
feat: Input&Output 에 해당하는 상세 이벤트들과 반환값 구현

- ViewController에서 받아오는 Event를 정의
- ViewModel 이 줄 정보들을 정의

July911 committed:
feat: LocationManager RxSwift 적용

- Observable<String>이 리턴되게 리팩토링

July911 committed:
feat: Coordinator에 RxSwift 적용

- DetailCityView의 City를 Observable로 적용
- DetailViewModel의 configureOutput private 메서드 생성

July911 committed:
refactor: 기존 ViewModel의 Delegate삭제

- ViewModel이 가지고 있었던 ViewModelDelegate삭제
- ViewController의 Extension에서 구현되었던 ViewModelDelegate 위임 제거
- ViewModelDelegate의 메서드 제거

July911 committed:
feat: DetailViewModel의 Output을 생성해주는 configureOutput메서드 생성

- ViewController가 ViewModel의 메서드를 직접 호출하는 부분 개선
- Output을 한곳에 모아서 코드의 가독성 증가
- WithUnretained를 해주어 self의 참조 문제 해결

July911 committed:
feat: DetailShowViewController Output&Input 바인딩

- 뷰에서 넘어오는 이벤트들을 받고 처리할 transform 상세 구현
- ViewModel이 주는 데이터를 TextView와 WebView에 바인딩

July911 committed:
feat: SearchViewModel 내부 Input&Output 정의

- SearchViewController의 TableViewDelegate 및 DataSource 삭제

July911 committed:
feat: SearchViewController ViewModelBind 구현

- SearchViewController bindViewModel 구현
- City를 observable에서 behaviorSubject로 변경

July911 committed:
feat: ReactiveExtension을 사용하여 이미지 캐싱 구현

- DetailShowViewController에 UIImageView와 WKWebView를 base로한 Extension 구현
- Button Event와 imageCacheData를 하나의 시퀀스로 연결

July911 committed:
refactor: 중복 구현된 combineLastest 삭제

- DetailViewModel Input에서 captureEvent삭제

July911 committed:
feat: SearchViewController 내부 구조 변경

- DistinctUntilChanged 를 사용하여 searchBar의 text가 변경되면 reloadData
- DefaultCitySearchRepository 오류 수정
- Input Event를 받아오는 Input을 삭제하고 text만으로 변경

July911 committed:
fix: CitySearchRepository 로직 수정

- name filtering로직을 hasPrefix로 수정

July911 committed:
feat: SearchViewController navigationController title 추가

July911 committed:
refactor: 함수형 프로그래밍으로 리팩토링

- Repository에서 search를 했을 때, 원본 데이터를 손상시키지않고 return
- 원본 Observable과 filtered된 Observable을 ViewModel에서 합성
- ViewController가 비지니스로직을 가지지 않고 수동적으로 반응하게 변경

July911 committed:
chore: Mark주석 추가

July911 committed:
feat: TableView의 Cell이 선택되었을때 deselected되게 구현

- itemSeleted를 통해 indexPath를 받아온 후 구독

July911 committed:
feat: UI components 색 보정 및 AttributtedString 적용

July911 committed:
feat: Localization 구현

- 영어로 보이던 도시 이름을 한국어로 지역화
- 지역화에 따른 filter hasPrefix 로직을 한국어로 변경

July911 committed:
refactor: Cell Text 및 TextView Text 크기 조정

- Cell constraint의 constant 변경
- String Extension 파일 하나로 병합 및 MARK 주석을 통한 가독성 증가

July911 committed:
refactor: ImageCache를 view가 사라질때 한번 실행 하도록 변경

- DetailShowViewController 캡쳐 버튼 삭제
- ViewWillDisappear시 캡쳐하도록 methodInvoked로 구현
- 강제 언래핑 수정

jsim27 commented:
먼저 readme 수정할 부분이 보여서 말씀드립니다!
> 1️⃣ 코딩 컨벤션: Karma 스타일  

커밋 컨벤션 말씀하시는거죠?

jsim27 commented on ChildOfWeather/Data/Repository/DefaultWeatherRepository.swift:
```diff
@@ -1,31 +1,36 @@
-            case .failure(_):
-                completion(nil)
-            }
+        return Observable<TodayWeather>.create { emitter in
+            self.service.request(decodedType: WeatherInformation.self, requestType: request)
+                .subscribe { (result) in
+                switch result {
+                case .success(let weatherInformation):
+                    emitter.onNext(weatherInformation.toDomain())
+                case .failure(let error):
+                    emitter.onError(error)
+                }
+            }.disposed(by: self.bag)
+            
+            return Disposables.create()
```
스트림이 끊길 필요가 있었을까요? 
저번에 대화 나눴을 때, 에러처리를 별개로 할 필요성 때문에 이렇게 할 수밖에 없었다고 들었던 것 같은데, 다시 한번 생각해 주세요😃
```swift
import RxSwift

extension Int {
    func toString() -> String {
        self.description
    }
}

enum SampleError: String, Error {
    case error = "에러~"
}

func single() -> Single<Int> {
    return Single.create { emitter in
        emitter(.failure(SampleError.error))
        return Disposables.create()
    }
}

single()
    .asObservable()
    .map { $0.toString() }
    .subscribe { event in
        print(event)
    }
// error(error)
```

July911 replied:
```swift 
   func fetchWeatherInformation(
        cityName text: String) -> Observable<TodayWeather> 
    {
        let request: RequestType = .getWeatherFromCityName(city: text)
        
        return self.service.request(decodedType: WeatherInformation.self, requestType: request)
            .map { $0.toDomain() }.asObservable()
    }
``` 
말씀해주신대로 구현했습니다 . 예시까지 ...정말 감사합니다 ...!

jsim27 commented on ChildOfWeather/NetWork/APIService.swift:
```diff
@@ -9,55 +10,60 @@ enum APICallError: Error {
+            let request = URLRequest(url: url)
+            let task = URLSession.shared.dataTask(with: request) { data, URLResponse, error in
+                
+                guard error == nil
```
서버에서 보내주는 error 정보는 사용하지 않나요?

July911 replied:
에러가 발생하면 error가 존재한다는 에러를 보내도록 수정했습니다 ...! 감사합니다

jsim27 commented on ChildOfWeather/NetWork/APIService.swift:
```diff
@@ -9,55 +10,60 @@ enum APICallError: Error {
+                
+                guard error == nil
+                else {
+                    return
```
서버에서 에러를 보낸 경우에 왜 아무 처리도 하지 않나요?

jsim27 commented on ChildOfWeather/NetWork/APIService.swift:
```diff
@@ -9,55 +10,60 @@ enum APICallError: Error {
+                else {
+                    single(.failure(APICallError.errorExist))
+                    return
+                }
```
`APICallerError`에는 다양한 상황에 대해 정의해놓으셨는데 왜 모든 에러핸들링이 같은 에러케이스로 처리되고 있나요?

July911 replied:
각각의 케이스별로 에러케이스를 만들어 적용시켜주었습니다 !

jsim27 commented on ChildOfWeather/NetWork/APIService.swift:
```diff
@@ -9,55 +10,60 @@ enum APICallError: Error {
-                completion(.failure(APICallError.invalidResponse))
-                return
+            return Disposables.create {
+                task.cancel()
```
배웠습니다.. Dispose되면 cancel!! 너무 좋네용😍

jsim27 commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -1,91 +1,117 @@
             let weatherDescription = "일출은 오전\(sunrise)\n일몰은 오후\(sunset)\n최고 기온은  섭씨\(maxTemp)도\n최저 기온은 섭씨\(minTemp)도입니다."
-                
-            self?.delegate?.loadTodayDescription(weather: weatherDescription)
+            return Observable.just(weatherDescription)
```
`flatMap`과 `Observable.just`의 사용이 의미가 없어보입니다. 
`map`을 사용하는것과 정확히 같은 동작을 할 것 같아요.

덧붙여, 이 스트림에서 Single -> Observable -> Observable(just)로 연결하려고 하신 의도를 설명해 주세요. 단일 이벤트만 받아올것이라면 쭉 Single로 이어도 되지 않나요?

July911 replied:
쭉 싱글로 이어지게 구현했습니다
부가적으로 map을 사용해 동작하게 해주었습니다.
단 질문이 있습니다. 
single의 onError 이벤트는 구독시에만 발생하는데 어디서 에러 핸들링을 해줘야 할까가 고민인데 나무의 의견은 어떠신지요?

jsim27 commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -1,91 +1,117 @@
+    }
+    // MARK: - Open Method
+    func transform(input: Input) -> Output? {
+        guard let output = self.configureOutput()
```
input으로부터의 스트림 없이 뷰모델에서부터 Output 스트림을 시작하는 모습이네요!
`binding`하지 않고 메서드를 호출하는 것.. Rx답지 않은 방식 같아요.
본 함수의 이름 `transform`에 더 부합하는 방식이 있을 것 같네요. 현 상황은 `Input`을 `Output`으로 바꾸어주는 것이 아니라, `ViewDidLoad`이벤트까지 바인딩 없이 `transform`메서드로 받아버린 격이네요.
VC의 methodInvoked로부터 시작되는 스트림을 받아서 이어가는 방식으로 하지 않으신 이유가 있으신가요?

July911 replied:
이 부분에 대하여 의논하고 싶은게 있습니다! 
```swift
   input.capturedImage
            .withUnretained(self)
            .filter { _ in self.imageCacheUseCase.hasCacheExist(cityName: self.extractCity().name) == false }
            .subscribe(onNext: { (self, image) in
                self.imageCacheUseCase.setCache(object: image)
            }).disposed(by: self.bag)
        
        input.touchUpbackButton
            .withUnretained(self)
            .observe(on: MainScheduler.instance)
            .subscribe( onNext: { _ in
                self.coordinator.occuredViewEvent(with: .dismissDetailShowUIViewController)
            }).disposed(by: self.bag)
``` 
백버튼이 눌렸을때와 이미지가 넘어왔을때 캐싱을 해주는 작업은 다시 View로 보낼 input이 없는 이벤트이고, 구조상 ViewModel 이 Coordinator를 들고있어서 ViewModel 단에서 처리해주어야하는 작업같은데
Observable은 subscribe이 되어야만 의미가 있어서 ( 데이터를 넘겨주고 ) 어떤 방식으로 구현해야 Rx다운 방식일지 고민이 됩니다. 
혹시 나무는 어떻게 생각하시나요? 

PS ) ViewController 단으로 다시 보내주는 작업은 모두 합성해서 보냈습니다 ...!

jsim27 replied:
https://github.com/sergdort/CleanArchitectureRxSwift
저희가 자주 참고하는 위 프로젝트에서, 굳이 뷰모델 -> 뷰로 돌아갈 필요가 없는 스트림을 어떻게 처리하고 있는지 참고해 보세요. 저도 이번에 보고 큰 깨달음을 얻었습니다...
보시고 모르겠다면 다시한번 말씀해주세요!

jsim27 replied:
덧붙여, 본 코멘트의 본래 의도는 왜 ViewWillAppear 스트림으로부터 작업을 시작하지 않고 `configureOutput()`하셨는지가 궁금했던거였어요!

jsim27 commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -1,91 +1,117 @@
+        }
+        
+        input.capturedImage
+            .withUnretained(self)
```
https://minios.tistory.com/46

`withUnretained()`의 사용법과 주의점에 대해서 조금더 알아봅시다ㅎㅎ
(@forestjae 가 알려주셨어요)

July911 replied:
각 상황별로 withUnretained를 사용해주어야하는군요 .. 
한번 적어주면 순환참조가 해결되는줄 알았는데 꿀팁 감사합니다 ...!
@forestjae  @jsim27  !!!

jsim27 commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -1,91 +1,117 @@
+            return nil
+        }
+        
+        input.capturedImage
+            .withUnretained(self)
+            .filter { _ in self.imageCacheUseCase.hasCacheExist(cityName: self.extractCity().name) == false }
+            .subscribe(onNext: { (self, image) in
+                self.imageCacheUseCase.setCache(object: image)
+            }).disposed(by: self.bag)
+        
+        input.touchUpbackButton
+            .withUnretained(self)
+            .observe(on: MainScheduler.instance)
+            .subscribe( onNext: { _ in
+                self.coordinator.occuredViewEvent(with: .dismissDetailShowUIViewController)
+            }).disposed(by: self.bag)
+        
+        return output
```
`View`에서 시작된 스트림이 `ViewModel`에서 소비되고 dispose되면 dispose하는 의미가 있을까요?
뷰모델에서의 subscribe는 여러모로 존재하면 안된다고 생각합니다. 뷰모델이 stream의 시작점이 될 이유가 없기 때문입니다.
이 부분은 더 의논해봐도 좋을 것 같아요!

July911 replied:
위의 대답과 같은 의문을 나무도 품으셨군요 !! 
한번 의논해보아도 좋을 것 같아요 어떤 방식으로 ViewModel의 dispose를 막을 수 있을까요?

jsim27 replied:
위 코멘트의 답변으로 대신하겠습니다!

jsim27 commented on ChildOfWeather/Presentation/SearchView/ViewModel/SearchViewModel.swift:
```diff
@@ -1,39 +1,47 @@
+
+        let combined = Observable.combineLatest(entireCities, filteredCities) { (city, fil) -> [City] in
+            return fil.isEmpty ? city : fil
         }
```
까다로운 stream 합성인데 잘 생각하신 것 같아요! 다음과 같은 방법도 있을 것 같습니다. 무엇이 더 좋은 방법인지는 잘 모르겠지만 참고만 해 보세요😀
```swift
let combined = filteredCities
    .flatMap {
        $0.isEmpty ? entireCities : Observable.of($0)
    }
```

July911 replied:
코멘트 답변 하면서 많이 배워갑니다...  flatMap의 사용성이 조금 더 명확해진거같아요 
코드도 더 간결해지고 알아보기 쉬워진거같습니다. 감사합니다.

jsim27 replied:
이 부분은 정말로 뭐가 낫다는게 아니었습니다. 저는 flatMap이 더 익숙할 뿐이라서요ㅎㅎ

jsim27 commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -1,91 +1,117 @@
-    let city: City
+final class DetailShowViewModel {
+    // MARK: - Private Property
+    private let city: BehaviorSubject<City>
```
stream으로 이어져 있는 것도 아닌데, 굳이 BehaviorSubject로 할 필요가 있었나요?
그냥 변수로 바꾸어 보아도 전혀 문제없이 작동하는 것처럼 보이는데, 의도를 설명해 주세용

July911 replied:
변하지 않는 데이터인데 behaviorSubject로 만들어서 괜히 Rx에 대한 의존성만 높아진 것 같습니다.
RxSwift를 사용하는 이유가 나중에 생기는 데이터를 리턴값으로 받을 수 있다는 것 + 비동기 작업 시 용이함인데 놓치고 간 부분인 것 같습니다. 코멘트 감사합니다 !

jsim27 replied:
이유가 있다면 사용해도 무방하지만, 단순한 데이터의 반환만 일어나는 것 같아서 말씀드려봤네요😄

jsim27 commented on ChildOfWeather/Presentation/SearchView/ViewModel/SearchViewModel.swift:
```diff
@@ -1,39 +1,47 @@
     }
+   // MARK: - Nested Type
+    struct Input {
+        let viewWillAppear: Observable<Void>
```
왜 인풋 스트림이 사용되지 않고 있을까요?

jsim27 replied:
@July911

July911 replied:
```swift
let citiesCombinedInputEvent = combined.sample(input.viewWillAppear)
```
 이런식으로 sample로 사용해주었습니다 !! 
viewWillAppeaer 이벤트가 들어오고 나서야 되게요 !!

jsim27 commented on ChildOfWeather/Presentation/SearchView/ViewModel/SearchViewModel.swift:
```diff
@@ -1,39 +1,47 @@
+        
+        input.didSelectedCell.subscribe(onNext: { (city) in
+            self.coordinator.occuredViewEvent(with: .presentDetailShowUIViewController(cityName: city))
+        }).disposed(by: self.bag)
```
다른 파일에서도 언급해 드렸지만, subscribe의 주체는 view가 되는것이 맞는것 같아요. 뷰가 요청한 스트림이 비동기적으로 실행되는 도중에, 해당 뷰가 deinit되어 disposeBag이 작동할 때 해당 스트림이 더이상 작동하지 않도록 보장하려면요!
더불어, 뷰는 자신이 보낸 Input을 transform을 통해 disposable 형태로 항상 응답받아야 합니다. 이는 반환받을 정보의 유무와 무관합니다.

July911 replied:
```swift 
let capturedSuccess = input.capturedImage
            .withUnretained(self)
            .filter { _ in
                self.imageCacheUseCase.hasCacheExist(cityName: self.extractCity().name) == false }
            .do(onNext: { (self, image) in
                self.imageCacheUseCase.setCache(object: image)
            }).map { _ in }
        
        let dismiss = input.touchUpbackButton
            .withUnretained(self)
            .observe(on: MainScheduler.instance)
            .do(onNext: { _ in
                self.coordinator.occuredViewEvent(with: .dismissDetailShowUIViewController)
            }).map { _ in }
``` 
이런식으로 do 시 행동만 적어주고 ViewController 쪽에서 구독하게 변경하였습니다.
저도 나무 코멘트 듣고 이것저것 찾아보았는데, 단방향 바인딩이라는 목적성에 맞게 ViewController 에서 모든 binding 을 해주어야 한다고 생각했습니다.
 클린아키텍쳐로 2번째 프로젝트인데, 매번 새로운것을 배우네요 ㅎㅎ

jsim27 replied:
아하, 위 코멘트에 대한 답을 찾으셨군요ㅎㅎㅎ 네, 저도 do로 작업하고 Output으로 다시 뷰로 보내서 구독하는 방법을 생각했었습니다!
저도 덕분에 배웠어요ㅎㅎ

jsim27 commented on ChildOfWeather/Presentation/DetailView/View/DetailShowViewController.swift:
```diff
@@ -111,53 +105,95 @@ final class DetailShowUIViewController: UIViewController {
+                    return
+                }
+                let request = URLRequest(url: url)
+                DetailviewController.webView.load(request)
```
데이터를 가공하는 로직이 ViewController에 있는 것 같네요. ViewModel에서 가공하지 않으신 이유가 있으신가요?

July911 replied:
비지니스 ..로직은 아니라고 생각했는데 ViewModel과 ViewController사이를 고민해주었던 것 같아요.
하지만 ViewModel에 위치하는 것이 조금 더 수동적인 ViewController를 유지할 수 있는 것 같습니다. 감사합니다 !

jsim27 replied:
넵 비즈니스로직은 도메인 레이어에서 처리하는게 맞고, 이런 경우 뷰 간의 로직에서 데이터 가공이 필요하다면 그 역할은 뷰모델이 담당하는게 적합하다고 생각했습니다! 👍

jsim27 commented on ChildOfWeather/Presentation/DetailView/ViewModel/DetailShowViewModel.swift:
```diff
@@ -1,91 +1,117 @@
+            return nil
+        }
+        
+        return Observable.just(image)
```
`hasCacheExist`를 검사하는 것은 repository 혹은 useCase의 역할이고, 뷰모델은 결과만 받으면 된다고 생각하는데 줄라이 의견은 어떠신가요?
+ 비동기 작업이 아니라면, 꼭 Observable로 내보낼 이유가 있을까요?

July911 replied:
이 부분 로직 수정했습니다 ! 
```swift 
  func fetchImage(cityName: String) -> ImageCacheData? {
        guard hasCacheExist(cityName: cityName)
        else {
            return nil
        }
        return self.imageProvideRepository.fetchCache(key: cityName)
    }
    
    private func hasCacheExist(cityName: String) -> Bool {
        self.imageProvideRepository.fetchCache(key: cityName) == nil ? false : true
    }
``` 

이런식으로요 !!

jsim27 commented on ChildOfWeather/Data/Repository/DefaultWeatherRepository.swift:
```diff
@@ -1,31 +1,36 @@
+    func fetchURLFromLocation(locationAddress address: String) -> String {
         let type: RequestType = .getMapfromLocationInformation(location: address)
         
         return type.fullURL
```
`TodayWeather` 타입을 다루는 repository에 이 메서드가 있는 것이 어색해 보여요.

July911 replied:
LocationManager가 이 로직을 가져가게 했습니다 !

jsim27 commented on ChildOfWeather/Data/Repository/DefaultCitySearchRepository.swift:
```diff
@@ -1,27 +1,37 @@
 final class DefaultCitySearchRepository: CitySearchRepository {
     
-    private var assetData: [City]?
+    private let assetData = BehaviorRelay<[City]>(value: [])
```
`fetchCityList`가 비동기 작업이 아닌 것으로 보이는데, 이 원본 데이터가 BehaviorRelay일 필요가 있을까요?
이후 데이터가 전달되는 흐름 또한, 원본이 BehaviorRelay여야만 할 이유는 없어 보여요!

July911 replied:
수정했습니다! 굳이 흐름이 RxSwift를 사용하지 않아도 될 것 같습니다 .!!

jsim27 commented on ChildOfWeather/Presentation/DetailView/View/DetailShowViewController.swift:
```diff
@@ -111,53 +105,95 @@ final class DetailShowUIViewController: UIViewController {
-        self.webviewSnapshot { }
+}
+// MARK: - Reactive Extension 
+private extension Reactive where Base: WKWebView {
+    func takeSnapShot(city: City) -> Observable<ImageCacheData> {
+        return Observable<ImageCacheData>.create { emitter in
+            let configuration = WKSnapshotConfiguration()
+            base.takeSnapshot(with: configuration) { image, error in
+                guard error == nil
+                else {
+                    emitter.onError(APICallError.dataNotfetched)
+                    return
+                }
+                
+                guard let image = image
+                else {
+                    emitter.onError(APICallError.errorExist)
+                    return
+                }
+                let imageCacheData = ImageCacheData(key: city.name as NSString, value: image)
+                emitter.onNext(imageCacheData)
+                emitter.onCompleted()
+            }
+            return Disposables.create()
+        }
     }
 }
+
+private extension Reactive where Base: UIImageView {
+    func loadCacheView(webView: WKWebView) -> Binder<UIImage> {
+        return Binder(self.base) { ImageView, image in
+            webView.isHidden = true
+            base.isHidden = false
+            ImageView.image = image
+        }
+    }
+}
```
진짜 멋집니다.. 👏👏👏👏👏👏👏👏👏👏

jsim27 reviewed (commented):
고생 많으셨습니다 줄라이!

Rx와 클린 아키텍쳐 위주로만 보아서, 세부 구현 상 놓친 부분이 있을 수 있는데, 혹시 더 같이 이야기해보고 싶은 부분이 있으시다면 추가로 편하게 말씀해 주세요!

대부분의 경우에 Rx 관련해서도 언급할 부분 없이 너무 잘 해 주셔서.. 더 많은 코멘트 드려야 하는데 더이상 못찾겠네요ㅎㅎ 너무 고생하셨습니다!!

언급드린 부분들도 제 의견일 뿐이니 더 같이 논의해보면 좋을 것 같습니다! 
코멘트 & 수정 반영 부탁드려요!

---

## 구현한 부분
> Reactive Extension 으로 필요한 부분 구현

한 수 배웠습니다 멋져용!!

> Observable 병합을 통한 비지니스 로직 이동
Transform 메서드가 인터페이스가 되어서 ViewController가 이 메서드만 호출하고 그로 인해 바인딩이 되게 구현했습니다.

Input으로 들어온 이벤트가 Repository나 Seivice의 변경을 요청하거나 결과를 요청하고, flatMap을 통해 결과 Observable을 flatten하여 그 결과를 Output으로 내보내는 모습이 정석적인 RxSwift 사용이네요👍 Input Output 모델링으로 기능 파악도 쉽게 할 수 있었습니다!

> 함수형 프로그래밍 구현
SearchBar를 통해 search를 할 경우에 원본 데이터는 손상시키지않고 언제나 같은 값이 나오도록 함수형 프로그래밍으로 메서드를 구현해보았습니다!

이 부분에서 궁금한게 있는데요!
현재 예시를 들어주신 코드에서, `search(name: )`은 `assetData` 영향을 받고 있는 모습이네요. 만약 `assetData`가 변경된다면 이 함수는 input name에 대해서 항상 같은 결과를 반환하는 함수가 아니게 될텐데, 이 점은 어떻게 생각하시나요?

## 궁금한점
> ViewController의 로직
> SearchBar의 Text가 들어왔을 때 reloadData()
item이 seleted 되었을 때 deseletRow되는 로직이 ViewController 에 들어있는데 이 정도까지는 괜찮을지가 궁금합니다 !

-> 뷰모델의 데이터를 변경하거나 요청하지 않는 순수한 뷰 관련 로직이기에 상관없을 것 같습니다!

---
알고 계실수도 있지만, 앱 차원에서 의도하신 부분인지 궁금한게 있습니다!
1. 서치바에 검색어 입력 후 Finishediting하지 않고 뷰의 다른 부분을 터치하면 검색어가 사라지고 검색결과가 표시되지 않는 점
2. 디테일뷰에서 pop했을 때, 최근 검색 결과가 여전히 뷰에 남아있지만 서치바에는 해당 내용이 없는 점이 부자연스러워 보이는 점

이것도 개인적인 의견일 뿐입니다..
전보다 디자인이 훨씬 나아져서 참 보기 좋네요.
마무리까지 화이팅해봐요~~

jsim27 commented on ChildOfWeather/Presentation/SearchView/View/SearchViewController.swift:
```diff
@@ -48,57 +61,56 @@ final class SearchViewController: UIViewController {
-        let cities = self.viewModel?.filterdResults
-        cell.configureCell(city: cities?[indexPath.row] ?? City.EMPTY)
+        self.navigationItem.searchController?.searchBar.rx.text
+            .distinctUntilChanged()
+            .subscribe(onNext: { event in
+                self.listTableView.reloadData()
+            }).disposed(by: self.bag)
```
이미 데이터가 바인딩되어 있는데 다시 reload할 필요가 있을까요?
제가 이 로직을 제거하고 실행해 봤을때는 정상작동하는것 같더라구요!

July911 replied:
민망하네요 ... ㅎㅎㅎ drive로 걸어주었는데 잊었던 것 같습니다 ㅠㅠ..

July911 committed:
refactor: APIRequest 에러case 상세 분리

July911 committed:
refactor: Observable로 이어주던 시퀀스를 Single로 변경

July911 committed:
refactor: SearchViewModel filtering 로직 수정

- SearchViewModel 의 Input에 viewWillAppear 메서드 추가

July911 committed:
refactor: RxSwift가 필요하지 않은 부분에 Observable 제거

- BehaviorRelay<[City]> 를 [City]로 변경

July911 committed:
refactor: ViewModel의 disposeBag을 파라미터로 받아 dispose 되게 변경

- DetailViewModel Nested Struct 내부 Input의 ViewWillAppear Event 추가
- ViewWillAppear 이벤트와 Output을 합성하여 View로 전달하게 구조 변경

July911 committed:
refactor: ViewModel에 있는 구독을 모두 삭제 및 ViewController 로 이동

- View의 input을 합성하여 output으로 값을 리턴하게 구조 변경
- ViewModel의 disposeBag 제거
- ViewController가 모든 구독을 받게 구조 변경

July911 committed:
refactor: DetailShowViewController에 있던 데이터 가공로직을 ViewModel로 이동

- URLString을 URLRequest로 ViewModel에서 변경하도록 수정

July911 committed:
refactor: CacheImage가 있는지 검증하는 로직을 ViewModel에서 UseCase로 이동

July911 committed:
refactor: Address를 URL로 바꿔주는 로직을 LocationManager 공용 인스턴스로 이동

July911 committed:
chore: 불필요한 인덴트 제거 및 띄어쓰기 수정

July911 committed:
refactor: DefaultCitySearchRepository RxSwift제거

- 불필요한 Observable 제거
- Merge 병합으로 viewWillAppear 이벤트와 병합
